File: google.py
import os
import random
import requests
from bs4 import BeautifulSoup
import argparse

class GoogleOSINTScraper:
    def __init__(self, query, num_pages):
        self.query = query
        self.num_pages = num_pages
        self.base_url = 'https://www.google.com/search?q='
        self.headers = {
            'User-Agent': '<YOUR USERAGENT>'
        }
        self.results = []

    def _selectors(self, element):
        '''Returns the appropriate CSS selector.'''
        selectors = {
            'links': 'div.g',
            'next': 'a#pnnext'
        }
        return selectors[element]

    def _parse_page(self, html):
        soup = BeautifulSoup(html, 'html.parser')
        links = soup.select(self._selectors('links'))
        for link in links:
            anchor = link.select_one('a')
            title_element = link.select_one('h3')
            if anchor and title_element:
                url = anchor.get('href')
                title = title_element.get_text()
                self.results.append({'url': url, 'title': title})


    def _save_to_file(self):
        output_file = f'output-{random.randint(1, 100000)}.txt'
        with open(output_file, 'w', encoding='utf-8') as file:
            for result in self.results:
                file.write(f"Title: {result['title']}\nURL: {result['url']}\n\n")

    def scrape(self):
        for page in range(1, self.num_pages + 1):
            url = f'{self.base_url}{self.query}&start={(page-1)*10}&num=100&hl=en'
            response = requests.get(url, headers=self.headers)

            if response.status_code == 200:
                self._parse_page(response.text)
                self._save_to_file()

                next_page_link = BeautifulSoup(response.text, 'html.parser').select_one(self._selectors('next'))

                if next_page_link:
                    next_page_url = f'https://www.google.com{next_page_link.get("href")}'
                    response = requests.get(next_page_url, headers=self.headers)
                else:
                    break
            else:
                print(f"Failed to retrieve page {page}. Status Code: {response.status_code}")
                break

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Contoh program dengan argparse')
    parser.add_argument('-s', '--search', help='search dorking', required=True)
    parser.add_argument('-p', '--pages', help='pages next', type=int, required=True)
    args = parser.parse_args()

    query = args.search
    num_pages = args.pages
    
    scraper = GoogleOSINTScraper(query, num_pages)
    scraper.scrape()

File: google.js
const fs = require('fs');
const axios = require('axios');
const cheerio = require('cheerio');
const argparse = require('argparse');

class GoogleOSINTScraper {
    constructor(query, numPages) {
        this.query = query;
        this.numPages = numPages;
        this.baseUrl = 'https://www.google.com/search?q=';
        this.headers = {
            'User-Agent': '<YOUR USERAGENT>'
        };
        this.results = [];
    }

    _selectors(element) {
        const selectors = {
            'links': 'div.g',
            'next': 'a#pnnext'
        };
        return selectors[element];
    }

    _parsePage(html) {
        const $ = cheerio.load(html);
        const links = $(this._selectors('links'));
        links.each((index, element) => {
            const anchor = $(element).find('a');
            const titleElement = $(element).find('h3');
            if (anchor && titleElement) {
                const url = anchor.attr('href');
                const title = titleElement.text();
                this.results.push({ 'url': url, 'title': title });
            }
        });
    }

    _saveToFile() {
        const currentDate = new Date();
        const timestamp = currentDate.getTime();
        const outputFile = `output-${timestamp}.txt`;
        const fileContent = this.results.map(result => `Title: ${result.title}\nURL: ${result.url}\n\n`).join('');
        fs.writeFileSync(outputFile, fileContent, 'utf-8');
    }

    async scrape() {
        for (let page = 1; page <= this.numPages; page++) {
            const url = `${this.baseUrl}${this.query}&start=${(page - 1) * 10}&num=100&hl=en`;
            try {
                const response = await axios.get(url, { headers: this.headers });
                if (response.status === 200) {
                    this._parsePage(response.data);
                    this._saveToFile();

                    const $ = cheerio.load(response.data);
                    const nextPageLink = $(this._selectors('next')).attr('href');

                    if (nextPageLink) {
                        const nextPageUrl = `https://www.google.com${nextPageLink}`;
                        const nextPageResponse = await axios.get(nextPageUrl, { headers: this.headers });
                    } else {
                        break;
                    }
                } else {
                    console.log(`Failed to retrieve page ${page}. Status Code: ${response.status}`);
                    break;
                }
            } catch (error) {
                console.error(`Error: ${error.message}`);
                break;
            }
        }
    }
}

const parser = new argparse.ArgumentParser({ description: 'Contoh program dengan argparse' });
parser.add_argument('-s', '--search', { help: 'search dorking', required: true });
parser.add_argument('-p', '--pages', { help: 'pages next', type: 'int', required: true });
const args = parser.parse_args();

const query = args.search;
const numPages = args.pages;

const scraper = new GoogleOSINTScraper(query, numPages);
scraper.scrape();
